src/storage/store/rel_table.cpp

```diff
@@ -49,10 +50,107 @@ vector<AdjColumn*> RelTable::getAdjColumnsForNodeTable(table_id_t tableID) {
     return retVal;
 }
 
+// Prepares all the db file changes necessary to update the "persistent" store of lists with the
+// listsUpdateStore, which stores the updates by the write trx locally.
 void RelTable::prepareCommitOrRollbackIfNecessary(bool isCommit) {
-    performOpOnListsWithUpdates(
-        std::bind(&Lists::prepareCommitOrRollbackIfNecessary, std::placeholders::_1, isCommit),
-        std::bind(&RelTable::addToUpdatedRelTables, this));
+    auto& listUpdatesPerDirection = listsUpdateStore->getListUpdatesPerTablePerDirection();
+    for (auto& relDirection : REL_DIRECTIONS) {
+        for (auto& listUpdatesPerTable : listUpdatesPerDirection[relDirection]) {
+            if (isCommit && !listUpdatesPerTable.second.empty()) {
+                auto srcTableID = listUpdatesPerTable.first;
+                auto listsUpdateIterators = getListsUpdateIterators(relDirection, srcTableID);
+                // Note: In C++ iterating through maps happens in non-descending order of the keys.
+                // This property is critical when using listsUpdateIterator, which requires the user
+                // to make calls to writeInMemListToListPages in ascending order of nodeOffsets.
+                auto& listUpdatesPerChunk = listUpdatesPerTable.second;
+                for (auto updatedChunkItr = listUpdatesPerChunk.begin();
+                     updatedChunkItr != listUpdatesPerChunk.end(); ++updatedChunkItr) {
+                    for (auto updatedNodeOffsetItr = updatedChunkItr->second.begin();
+                         updatedNodeOffsetItr != updatedChunkItr->second.end();
+                         updatedNodeOffsetItr++) {
+                        auto nodeOffset = updatedNodeOffsetItr->first;
+                        auto& listUpdates = updatedNodeOffsetItr->second;
+                        // Note: An empty listUpdates can exist for a nodeOffset, because we don't
+                        // fix the listUpdates, listUpdatesPerNode and ListUpdatesPerChunk indices
+                        // after we insert or delete a rel. For example: a user inserts 1 rel to
+                        // nodeOffset1, and then deletes that rel. We will end up getting an empty
+                        // listUpdates for nodeOffset1.
+                        if (!listUpdates.hasUpdates()) {
+                            continue;
+                        }
+                        if (listUpdates.emptyListInPersistentStore) {
+                            listsUpdateIterators[0]->updateList(nodeOffset,
```

Comment:
As a second comment I'm now debating if this first `if clause` is justified. Why don't you just have 2 cases: 

1) if the list is a large list and there are only updates;
2) else

So you don't differentiate between whether the persistent list was empty or not and just call iterator->updateList(...), which will check whether to fill from the persistent store or not. I'm assuming updateList() should actually just work even if the persistent store is empty. It would simplify the code.

Given that you do this, I have a follow up question: Aside from here do we ever use emptyListInPersistentStore? If we remove this if case, it may be unused and if so, we should remove it too.

Reply:
As we discussed, let's keep the `listsUpdates.emptyListInPersistentStore` branch here since the logic is quite different from the newly inserted nodes and other nodes:
```
If the node is newly inserted, we should only read from the updateStore.
If the node is not newly inserted, we should both read from the persistent and update store.
```
btw: renamed the `emptyListInPersistentStore` -> `newlyAddedNode`